### PR TITLE
Iter5

### PR DESCRIPTION
**Pull request by mrkovshik**

### CODE REVIEW

File: api/handler.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
+package api
```
Comment (arttet) [inline api/handler.go:1]: NICE TO HAVE: it would be better if you put a swagger file here.

File: api/handler.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"github.com/go-chi/chi/v5"
+	"github.com/mrkovshik/yametrics/internal/metrics"
+	service "github.com/mrkovshik/yametrics/internal/service/server"
+	"github.com/mrkovshik/yametrics/internal/storage/server"
+	"net/http"
+	"strconv"
+)
+
+func UpdateMetric(s *service.Server) func(http.ResponseWriter, *http.Request) {
```
Comment (arttet) [inline api/handler.go]: CHANGE REQUEST: please don't use anonymous functions

File: api/handler.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	return func(res http.ResponseWriter, req *http.Request) {
+		metricName := chi.URLParam(req, "name")
+		metricValue := chi.URLParam(req, "value")
+		metricType := chi.URLParam(req, "type")
+		switch metricType {
+
+		case metrics.MetricTypeGauge:
```
Comment (arttet) [inline api/handler.go]: CHANGE REQUEST: business logic is here. Please split your application by layers.

File: cmd/agent/agent.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package main
+
+import (
+	"fmt"
+	config "github.com/mrkovshik/yametrics/internal/config/agent"
```
Comment (arttet) [inline cmd/agent/agent.go:7]: SUGGESTION: please sort imports https://stackoverflow.com/questions/38704769/golang-import-grouping-by-package

File: internal/metrics/metric.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+type Imetric interface { //TODO: Подумать, нужен ли тут интерфейс вообще
+	Update(server.IStorage) error
+}
+
+const (
+	MetricTypeGauge   = "gauge"
+	MetricTypeCounter = "counter"
+)
+
+var MetricNamesMap = map[string]struct{}{
```
Comment (arttet) [inline internal/metrics/metric.go]: CHANGE REQUEST: please don't use global variables

File: internal/metrics/mock.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"fmt"
+	storage "github.com/mrkovshik/yametrics/internal/storage/agent"
+	"log"
+	"math/rand"
+	"time"
+)
+
+type MockMetrics struct {
```
Comment (arttet) [inline internal/metrics/mock.go:12]: SUGGESTION:
https://github.com/uber-go/mock

File: internal/storage/server/map_storage.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	value, ok := m.Gauges[name]
+	if !ok {
+		return "", errors.New("not found")
+	}
+	return fmt.Sprint(value), nil
+}
+
+func (m *MapStorage) GetAllMetrics() string {
+
+	resp := "<html><body><h1>Metric List</h1>" +
```
Comment (arttet) [inline internal/storage/server/map_storage.go]: CHANGE REQUEST: please use `strings.Builder`
https://stackoverflow.com/questions/1760757/how-to-efficiently-concatenate-strings-in-go

**Review by arttet — CHANGES_REQUESTED** [review]
Please add
- https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- https://github.com/golangci/golangci-lint

File: api/handler.go (hunk 2 above)
Comment (mrkovshik) [reply api/handler.go]: why not? it seems like a common practice to use such constructions to make an enclosure. Is it some kind of specific condition for this project?

File: api/handler.go (hunk 3 above)
Comment (mrkovshik) [reply api/handler.go]: moved this part to storage package

File: internal/metrics/metric.go (hunk 5 above)
Comment (mrkovshik) [reply internal/metrics/metric.go]: I moved this map inside of the SendMetric method, but I'm not quite sure if this is the best solution. I feel like it should be available all over the project (however it forks just fine now).

File: internal/storage/server/map_storage.go (hunk 7 above)
Comment (mrkovshik) [reply internal/storage/server/map_storage.go]: done

File: cmd/agent/agent.go (hunk 4 above)
Comment (mrkovshik) [reply cmd/agent/agent.go:7]: Sorted imports with Goland utils

File: api/handler.go (hunk 2 above)
Comment (mrkovshik) [reply api/handler.go]: Refactored without anonymous funcs

File: api/handler.go (hunk 1 above)
Comment (mrkovshik) [reply api/handler.go:1]: added OpenAPI spec

File: internal/metrics/mock.go (hunk 6 above)
Comment (mrkovshik) [reply internal/metrics/mock.go:12]: I read gomock docs, and honestly, I don't see how is it going to work here, maybe i'm gonna need to discuss it on 1:1 meeting. Anyway, as long as this comment marked as "SUGGESTION", I guess I can leave it as is by now

File: .github/workflows/coveragetest.yml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 2
+      - uses: actions/setup-go@v2
+        with:
+          go-version: '1.18'
+      - name: Run coverage
+        run: go test ./... -race -coverprofile=coverage.out -covermode=atomic
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@v3
```
Comment (arttet) [inline .github/workflows/coveragetest.yml]:
```suggestion
        uses: codecov/codecov-action@v4
```

File: .github/workflows/coveragetest.yml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 2
+      - uses: actions/setup-go@v2
+        with:
+          go-version: '1.18'
```
Comment (arttet) [inline .github/workflows/coveragetest.yml]:
```suggestion
          go-version: '1.21'
```

File: .github/workflows/coveragetest.yml [hunk 10]
```diff
@@ -0,0 +1,18 @@
+name: Test and coverage
+
+on: [pull_request]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (arttet) [inline .github/workflows/coveragetest.yml]:
```suggestion
      - uses: actions/checkout@v4
```

File: internal/service/server/service.go [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+import (
+	"log"
+
+	config "github.com/mrkovshik/yametrics/internal/config/server"
+	"github.com/mrkovshik/yametrics/internal/storage/server"
+)
+
+type Server struct {
+	Storage storage.IServerStorage
+	Logger  *log.Logger
```
Comment (arttet) [inline internal/service/server/service.go]: NICE TO HAVE: usually, `logger` is the last parameter.

File: internal/storage/agent/map_storage.go [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+func (m *AgentMapStorage) UpdateCounter() error {
+	var (
+		intValue int
+		err      error
+	)
+
+	value, ok := m.Map.Load("PollCount")
+	if !ok {
+		intValue = 0
```
Comment (arttet) [inline internal/storage/agent/map_storage.go]: SUGGESTION: it does not make sense.

File: internal/storage/server/map_storage.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+		if !ok {
+			return "", errors.New("not found")
+		}
+		stringValue = fmt.Sprint(value)
+	}
+	return stringValue, nil
+}
+
+func (s *MapStorage) GetAllMetrics() string {
+	var sb strings.Builder
```
Comment (arttet) [inline internal/storage/server/map_storage.go]: SUGGESTION: it should be a template
https://pkg.go.dev/text/template

**Review by arttet — APPROVED** [review]
Good job!

File: internal/storage/server/map_storage.go (hunk 13 above)
Comment (mrkovshik) [reply internal/storage/server/map_storage.go]: updated

File: internal/storage/agent/map_storage.go (hunk 12 above)
Comment (mrkovshik) [reply internal/storage/agent/map_storage.go]: updated

File: internal/service/server/service.go (hunk 11 above)
Comment (mrkovshik) [reply internal/service/server/service.go]: updated

File: .github/workflows/coveragetest.yml (hunk 10 above)
Comment (mrkovshik) [reply .github/workflows/coveragetest.yml]: updated

File: .github/workflows/coveragetest.yml (hunk 9 above)
Comment (mrkovshik) [reply .github/workflows/coveragetest.yml]: updated

File: .github/workflows/coveragetest.yml (hunk 8 above)
Comment (mrkovshik) [reply .github/workflows/coveragetest.yml]: updated

**Pull request merged**